Src/Support/Google.Apis.Core/Http/HttpClientFromMessageHandlerFactory.cs

```diff
@@ -0,0 +1,198 @@
+                // Simple: nothing will be decompressing content, because we have a configured handler that does not
+                // perform automatic decompression, so we can just intercept the original handler.
+                return new StreamInterceptionHandler(configuredMessageHandler.MessageHandler);
+            }
+            else if (configuredMessageHandler.PerformsAutomaticDecompression)
+            {
+                // Complex: we want to use a simple handler with no interception but with built-in decompression
+                // for requests that wouldn't perform interception anyway, and a longer chain for interception cases.
+
+                var noCompressionMessageHandler = DisallowDecompressionOptions.CheckItMatches(_httpMessageHandlerFactory(DisallowDecompressionOptions));
+
+                return new TwoWayDelegatingHandler(
+                    // Normal handler (with built-in decompression) when there's no interception.
+                    configuredMessageHandler.MessageHandler,
+                    // Alternative handler for requests that might be intercepted, and need that interception to happen
+                    // before decompression.
+                    new GzipDeflateHandler(new StreamInterceptionHandler(noCompressionMessageHandler.MessageHandler)),
+                    request => StreamInterceptionHandler.GetInterceptorProvider(request) != null);
+            }
+            else
+            {
+                // Simple: we have to create our own decompression handler anyway, so there's still just a single chain.
+                var interceptionHandler = new StreamInterceptionHandler(configuredMessageHandler.MessageHandler);
+                return new GzipDeflateHandler(interceptionHandler);
+            }
+        }
+
+        /// <summary>
+        /// Specifies the configuration options for a message handler.
+        /// </summary>
+        public class HttpMessageHandlerOptions
+        {
+            /// <summary>
+            /// Whether the message handler built from these options
+            /// may perform automatic decompression or not.
+            /// If set to true, the message handler may or may not perform automatic decompression.
+            /// If set to false, the message handler must not perform automatic decompression.
+            /// </summary>
+            public bool MayPerformDecompression { get; }
+
+            /// <summary>
+            /// Whether the message handler built from these options
+            /// may handle redirects or not. Redirects that are not handled
+            /// should buble up the handlers chain.
+            /// If set to true, the message handler may or may not handle redirects.
+            /// If set to false, the message handler must not handle redirects.
+            /// </summary>
+            public bool MayHandleRedirects { get; }
+
+            internal HttpMessageHandlerOptions(bool mayPerformDecompression, bool mayHandleRedirects)
```

Comment:
What's going to call this constructor? As far as I can see, we don't actually use the redirection handling now... Is this effectively left over from the previous approach?

Reply:
(Or more to the point, is there logic missing here?)